franklin-crypto/src/circuit/rescue.rs

```diff
@@ -0,0 +1,191 @@
+extern crate num_bigint;
+use self::num_bigint::{BigUint,ToBigUint};
+use std::str::FromStr;
+use super::uint32::UInt32;
+use super::multieq::MultiEq;
+use super::boolean::Boolean;
+use bellman::{ConstraintSystem, SynthesisError};
+use bellman::pairing::Engine;
+use circuit::Assignment;
+use bellman::pairing::ff::{
+    Field,
+    SqrtField,
+    PrimeField,
+    PrimeFieldRepr,
+    BitIterator
+};
+use bellman::{
+    LinearCombination,
+    Variable
+};
+use super::num::AllocatedNum;
+
+
+fn raise_to_power<CS,E>(
+    mut cs: CS,
+    x: &AllocatedNum<E>,
+    alpha: &E::Fr
+) -> Result<AllocatedNum<E>, SynthesisError>
+    where CS: ConstraintSystem<E>,E:Engine
+{
+        assert_ne!(*alpha,E::Fr::zero());
+        assert_ne!(*alpha,E::Fr::one());
+        let alpha_bits_be = {
+            let mut tmp = Vec::with_capacity(E::Fr::NUM_BITS as usize);
+            let mut found_one = false;
+            for b in BitIterator::new(alpha.into_repr()) {
+                found_one |= b;
+                if !found_one {
+                    continue;
+                }
+                tmp.push(b);
+            }
+            tmp
+        };
+        let squares={//define square, square of square, etc variables
+            let mut tmp:Vec<AllocatedNum<E>>=vec![x.clone()];
+            //alpha is constant
+            for i in 1..alpha_bits_be.len(){
+                let sqr=tmp.last().unwrap().square(cs.namespace(|| format!("sqr{}",i)))?;
+                tmp.push(sqr);
+            }
+            tmp
+        };
+        assert_eq!(squares.len(),alpha_bits_be.len());
+        let res={
+            let n=alpha_bits_be.len();
+            let mut tmp=squares[n-1].clone();
+            //alpha is constant
+            for i in 1..n{
+                if alpha_bits_be[i] {
+                    tmp=squares[n-1-i].mul(cs.namespace(|| format!("mul_due_to_bit{}",i)),&tmp)?;
+                }
+            }
+            tmp.clone()
+        };
+        Ok(res) 
+}
+
+fn fr_to_biguint<Fr: PrimeField>(fr: &Fr) -> BigUint {
+    let mut buffer = Vec::<u8>::new();
+    Fr::char()
+        .write_be(&mut buffer)
+        .expect("failed to write into Vec<u8>");
+    buffer.clear();
+    fr.into_repr()
+        .write_be(&mut buffer)
+        .expect("failed to write into Vec<u8>");
+    let value = BigUint::from_bytes_be(&buffer);
+    value
+}
+
+fn biguint_to_fr<F: PrimeField>(bigint: &BigUint) -> Option<F> {
+    F::from_str(&bigint.to_str_radix(10))
+}
+
+fn get_alphas<F:PrimeField>() -> Result<(F,F), SynthesisError>{
+    let zero=BigUint::from_str("0").unwrap();
+    let one=BigUint::from_str("1").unwrap();
+    let p_m_1={
+        let mut p=F::char();
+        p.sub_noborrow(&F::one().into_repr());
+        let p=F::from_repr(p).unwrap();
+        fr_to_biguint(&p)
+    };
+    let alpha={
+        let mut erath=vec![BigUint::from_str("2").unwrap(),BigUint::from_str("3").unwrap()];
+        let mut res=erath.last().unwrap().clone();
```

Comment:
Check for GCD instead. Take candidate alphas as input to the function and don’t care about their primality. Alpha is highly unlikely to be more than 7 for reasonable efficiency

Reply:
if alpha is prime, then GCD(alpha, p-1)==1 <==> (p-1) % alpha != 0. alpha has no other divisors, and i check this condition